torch/utils/checkpoint.py

```diff
@@ -37,17 +57,32 @@ def backward(ctx, *args):
         if not torch.autograd._is_checkpoint_valid():
             raise RuntimeError("Checkpointing is not compatible with .grad(), please use .backward() if possible")
         inputs = ctx.saved_tensors
+        if ctx.preserve_rng_state:
+            # Stash the surrounding rng state, and mimic the state that was
+            # present at this time during forward.  If cuda was not initialized
+            # at this time during the original forward, we assume that the imminent
+            # forward-segment-within-backward will not alter the cuda rng state.
+            current_cpu_rng_state = torch.get_rng_state()
+            torch.set_rng_state(ctx.fwd_cpu_rng_state)
+            if ctx.had_cuda_in_fwd:
+                current_cuda_rng_state = torch.cuda.get_rng_state()
+                torch.cuda.set_rng_state(ctx.fwd_cuda_rng_state)
         detached_inputs = detach_variable(inputs)
         with torch.enable_grad():
             outputs = ctx.run_function(*detached_inputs)
+        if ctx.preserve_rng_state:
+            # Restore the surrounding rng state
+            torch.set_rng_state(current_cpu_rng_state)
+            if ctx.had_cuda_in_fwd:
+                torch.cuda.set_rng_state(current_cuda_rng_state)
 
         if isinstance(outputs, torch.Tensor):
             outputs = (outputs,)
         torch.autograd.backward(outputs, args)
-        return (None,) + tuple(inp.grad for inp in detached_inputs)
+        return (None, None) + tuple(inp.grad for inp in detached_inputs)
 
 
-def checkpoint(function, *args):
+def checkpoint(function, *args, preserve_rng_state=False):
```

Comment:
We should really default to `True` :confused:

Reply:
I've moved `preserve_rng_state` to a global flag, defaulted to True.  This feels cleaner because we don't need to pass the additional arg around (which python 2.7 wasn't allowing anyway).